### PR TITLE
Type safety - Incremental step 2

### PR DESCRIPTION
**Pull request by Tcharl**
More type safety on JDL

---

Please make sure the below checklist is followed for Pull Requests.

- [x] [All continuous integration tests](https://github.com/jhipster/generator-jhipster/actions) are green
- [x] Tests are added where necessary
- [x] The JDL part is updated if necessary
- [x] [jhipster-online](https://github.com/jhipster/jhipster-online) is updated if necessary
- [x] Documentation is added/updated where necessary
- [x] Coding Rules & Commit Guidelines as per our [CONTRIBUTING.md document](https://github.com/jhipster/generator-jhipster/blob/main/CONTRIBUTING.md) are followed

When you are still working on the PR, consider converting it to Draft (below reviewers) and adding `skip-ci` label, you can still see CI build result at your branch.

### CODE REVIEW

File: jdl/converters/jdl-to-json/jdl-with-applications-to-json-converter.spec.ts [hunk 1]
```diff
@@ -1110,6 +1110,8 @@ JSONEntity {
                   global: {
                     custom: 42,
                   },
+                  source: {},
```
Comment (Tcharl) [inline jdl/converters/jdl-to-json/jdl-with-applications-to-json-converter.spec.ts:1113]: The use of a record in the convert method forces to have the 3 defined. wouldn't it be better to be able to specify a partial here?

File: jdl/converters/types.ts [hunk 2]
```diff
@@ -1,21 +1,26 @@
...
+export type JSONRelationship = {
   relationshipSide?: RelationshipSide;
   relationshipName: string;
   relationshipType: RelationshipType;
   otherEntityName: string;
   options?: Record<string, boolean | string | number>;
 } & Record<string, any>;
 
-export type Entity = {
+export type JSONEntity = {
```
Comment (Tcharl) [inline jdl/converters/types.ts:17]: I'm a bit confused with this one: what's the difference with `generators/base-application/types/entity.d.ts`? is it relevant that these `converters/types` exists?

File: jdl/exporters/jdl-exporter.spec.ts [hunk 3]
```diff
@@ -103,7 +104,10 @@ describe('jdl - JDLExporter', () => {
             let jdlObject;
             beforeEach(() => {
               jdlObject = new JDLObject();
-              jdlObject.addApplication(new JDLApplication({ config: { [CLIENT_FRAMEWORK]: NO_CLIENT_FRAMEWORK } }));
+              const jdlApplication: JSONApplication = {
```
Comment (Tcharl) [inline jdl/exporters/jdl-exporter.spec.ts:107]: Same confusion about the relevance: shouldn't it be `jdl/models/jdl-application.ts`

File: jdl/models/jdl-application.ts [hunk 4]
```diff
@@ -24,12 +24,18 @@ import JDLApplicationConfigurationOption from './jdl-application-configuration-o
 import JDLApplicationConfiguration from './jdl-application-configuration.js';
 import JDLApplicationEntities from './jdl-application-entities.js';
 import JDLOptions from './jdl-options.js';
+import AbstractJDLOption from './abstract-jdl-option.js';
+
+export type JDLJSONApplicationConfiguration = {
```
Comment (Tcharl) [inline jdl/models/jdl-application.ts:29]: Same confusion about the relevance: shouldn't it be jdl/models/jdl-application.ts

**Review by mshima — APPROVED** [review]

File: jdl/converters/types.ts (hunk 2 above)
Comment (mshima) [reply jdl/converters/types.ts:17]: `generators/base-application/types/entity.d.ts` have calculated values and supported annotations.
They are used in generators.
In jdl and json only base data and annotations of type `Record<any, any>` exists.

**Pull request merged**